openapi-service/modules/openapi-service-ls-extension/src/main/java/io/ballerina/openapi/extension/OpenAPIConverterService.java

```diff
@@ -0,0 +1,250 @@
+                        Path.of(request.getDocumentFilePath()), request.isEnableBalExtension());
+                //Response should handle
+                if (!yamlContent.isEmpty() && (yamlContent.get(0).getOpenAPI().isPresent())) {
+                    Optional<String> yaml = yamlContent.get(0).getYaml();
+                    yaml.ifPresent(response::setYamlContent);
+                } else {
+                    response.setError("Error occurred while generating yaml.");
+                }
+            }
+            return response;
+        });
+    }
+
+    /**
+     * This API is used to return the Json Array of {@code OASResult} that containing all the service node generated
+     * yaml content, json content, openapi content and all diagnostics and error message if generation failed.
+     */
+    @JsonRequest
+    public CompletableFuture<OpenAPIConverterResponse> generateOpenAPI(OpenAPIConverterRequest request) {
+        return CompletableFuture.supplyAsync(() -> {
+            OpenAPIConverterResponse response = new OpenAPIConverterResponse();
+            Path filePath = Path.of(request.getDocumentFilePath());
+            Optional<SemanticModel> semanticModel = workspaceManager.semanticModel(filePath);
+            Optional<Project> project = workspaceManager.project(filePath);
+            if (semanticModel.isEmpty()) {
+                response.setError("Error while generating semantic model.");
+                return response;
+            }
+            if (project.isEmpty()) {
+                response.setError("Error while getting the project.");
+                return response;
+            }
+            project = Optional.of(project.get().duplicate());
+            DiagnosticResult diagnosticsFromCodeGenAndModify = project.get()
+                    .currentPackage()
+                    .runCodeGenAndModifyPlugins();
+            boolean hasErrorsFromCodeGenAndModify = diagnosticsFromCodeGenAndModify.diagnostics().stream()
+                    .anyMatch(d -> DiagnosticSeverity.ERROR.equals(d.diagnosticInfo().severity()));
+            Collection<Diagnostic> compilationDiagnostics = project.get().currentPackage()
+                    .getCompilation().diagnosticResult().diagnostics();
+            boolean hasCompilationErrors = compilationDiagnostics.stream()
+                    .anyMatch(d -> DiagnosticSeverity.ERROR.equals(d.diagnosticInfo().severity()));
+            if (hasCompilationErrors || hasErrorsFromCodeGenAndModify) {
+                // if there are any compilation errors, do not proceed
+                response.setError("Given Ballerina file contains compilation error(s).");
+                return response;
+            }
+            Module defaultModule = project.get().currentPackage().getDefaultModule();
+            SemanticModel updatedSemanticModel =
+                    project.get().currentPackage().getCompilation().getSemanticModel(defaultModule.moduleId());
```

Comment:
Not safe to get the compliation without a lock. Use `workspaceManager.waitAndGetPackageCompilation` instead